README.md
# About
RustMan is a tool I developed based on the RCON interface of Rust servers (that's Rust the game by Facepunch). Instead of having to use binary modifications that would cause servers to become "modded" and adding way too much functionality, this mod can run on a completely different machine than the Rust server and does nothing Rust can't do as it only uses RCON commands and therefore already implemented mechanisms.

This version RustMan-php is the third iteration of an undescribable mess that I caused in PHP, I'm sorry. I will never again make such a large standalone program in PHP. The first one was tiny and bad and all crammed into a single file, the second one tried to resolve a really bad error in Rust RCON but failed miserably and the third and final PHP iteration tried to make some stable-ish version. It's stable until the server has more than 42 players connected which is the point when RCON packages are split up and cause huge problems. There's a fix for this but I'm not going to implement it because you're not supposed to use this project.

# Why??
Well fuck you, that's why. This is not for use and if you do use it, it's your own fault and nobody will help you.
This project is obsolete by the C# iteration and also the fact that Rust has been remade completely and therefore most commands used in this program are invalid now. Also colours.

# Your code sucks!
I know, shut up. I never meant for this crap to be visible to any other person than me so you better be happy that you have the honor of looking at this shit code.

# Why PHP, are you retarded?
All this started by me making a working RCON class in PHP to query my server for information and put it on my website. I then made a proof-of-concept manager that would log chat and spawn the occasional random airdrop. It all exploded from there. I added shit all the time until it's become the fat blob that it is today.

# What are the specs? How does it work? Are there instructions?
No. If you want to know stuff about this, dig into the code and try to figure it out. I'm sorry if you do.


data/history.txt
1399156917	2
1399157287	1
1399157297	2
1399157637	1
1399158436	2
1399160466	1
1399161994	2
1399162296	3
1399164936	4
1399164976	3
1399166397	2
1399168377	1
1399169553	2
1399169717	1
1399172877	0
1399187528	1
1399187628	2
1399190478	1
1399201778	3
1399201848	4
1399201878	5
1399202938	6
1399204098	7
1399204218	6
1399204298	7
1399204698	8
1399204898	9
1399205178	8
1399205598	7
1399205698	8
1399206578	7
1399206588	8
1399209749	7
1399209759	6
1399209779	5
1399212119	6
1399212179	5
1399213139	4
1399213949	5
1399214029	6
1399215699	5
1399216039	6
1399216729	5
1399216899	4
1399217429	5
1399217459	6
1399217740	7
1399217859	6
1399217959	7
1399219009	6
1399219229	5
1399220659	6
1399222029	5
1399222879	6
1399223319	5
1399223729	4
1399223909	3
1399224689	2
1399224819	1
1399228480	2
1399230051	3
1399230202	4
1399230551	5
1399232241	4
1399232251	5
1399232281	4
1399232711	5
1399232861	6
1399232981	5
1399232991	6
1399233351	5
1399233369	4
1399233391	5
1399233731	4
1399233741	5
1399234021	4
1399234061	5
1399234331	4
1399235871	3
1399235881	4
1399236471	3
1399238571	4
1399238591	5
1399241231	6
1399241241	7
1399241601	6
1399241739	5
1399241780	4
1399242560	3
1399242620	2
1399242640	1
1399242700	2
1399244232	3
1399244322	2
1399244332	3
1399244352	2
1399245802	1
1399245812	2
1399246422	1
1399246482	0
1399257745	1
1399257804	0
1399270086	1
1399271976	0
1399275578	1
1399275588	0
1399276348	1
1399281339	0
1399283139	1
1399284534	0
1399284536	1
1399287296	2
1399288186	3
1399288666	2
1399288826	1
1399289236	2
1399290276	1
1399292686	0
1399293056	1
1399293136	2
1399293196	1
1399294016	0
1399299768	1
1399300048	0
1399300228	1
1399300588	0
1399300718	1
1399300898	0
1399301078	1
1399301508	2
1399302038	3
1399302168	4
1399302738	5
1399304458	4
1399304538	3
1399304638	4
1399305189	3
1399305209	4
1399305249	3
1399305259	4
1399306990	3
1399307000	4
1399307020	5
1399307080	4
1399307090	5
1399308110	6
1399309060	7
1399309190	6
1399309590	5
1399309670	4
1399309710	3
1399309720	4
1399309890	5
1399310120	4
1399310150	5
1399310410	6
1399310810	7
1399310870	6
1399310940	5
1399311050	6
1399311210	5
1399311240	6
1399311440	7
1399311460	6
1399311470	7
1399311970	6
1399312050	7
1399312560	8
1399312660	7
1399312830	8
1399312970	7
1399313060	8
1399313610	7
1399313650	8
1399313850	9
1399313920	10
1399313950	9
1399314100	10
1399314230	9
1399314430	8
1399314450	7
1399314500	6
1399314560	7
1399314680	8
1399314710	9
1399315060	8
1399315110	7
1399315230	6
1399315340	5
1399315530	4
1399315870	5
1399315920	7
1399316300	6
1399316380	5
1399316620	4
1399316720	3
1399317500	4
1399318280	5
1399320370	6
1399321992	5
1399322292	6
1399322512	7
1399322702	6
1399323172	7
1399323682	6
1399325483	7
1399325978	6
1399325980	7
1399326973	8
1399327003	7
1399327093	6
1399327343	7
1399327403	6
1399327413	7
1399327463	6
1399328233	5
1399328433	6
1399328523	5
1399328723	4
1399330133	2
1399330173	1
1399333314	2
1399333524	1
1399333534	2
1399333704	3
1399336744	2
1399337074	3
1399337754	2
1399338524	1
1399339254	0
1399339264	1
1399340054	0
1399344526	1
1399345606	0
1399345636	1
1399345816	0
1399356926	1
1399357506	0
1399360046	1
1399360946	2
1399361026	1
1399363707	0
1399371693	1
1399382934	3
1399384914	4
1399385394	5
1399387445	6
1399387656	5
1399388006	6
1399389126	7
1399389186	8
1399389436	7
1399389587	6
1399389867	5
1399389887	6
1399389939	5
1399389959	6
1399390037	5
1399390047	6
1399390059	5
1399390079	6
1399390210	5
1399390220	6
1399390267	5
1399390277	6
1399390302	5
1399390312	6
1399390412	5
1399390422	6
1399390542	5
1399390562	7
1399390652	6
1399390682	5
1399390692	6
1399390823	5
1399390843	6
1399390853	7
1399390973	6
1399390983	7
1399391123	6
1399391133	7
1399391184	6
1399391204	7
1399391367	8
1399391427	7
1399391437	8
1399391477	7
1399391497	8
1399391697	6
1399392088	5
1399392098	6
1399392328	7
1399392588	6
1399393018	5
1399393028	6
1399393108	5
1399393118	4
1399393128	5
1399393148	4
1399393248	5
1399393758	6
1399394028	7
1399394178	6
1399394188	7
1399394908	8
1399394968	7
1399394988	8
1399395078	7
1399395198	8
1399395728	9
1399395878	8
1399396328	9
1399396378	10
1399397078	9
1399397548	8
1399397558	9
1399398218	8
1399398228	9
1399398488	8
1399399249	9
1399399449	10
1399399639	9
1399399690	10
1399399939	11
1399400009	10
1399400019	11
1399400189	10
1399400199	11
1399400389	10
1399400489	11
1399400497	10
1399400509	11
1399400629	12
1399401469	11
1399401839	10
1399402309	9
1399402499	10
1399402649	11
1399402909	12
1399403279	11
1399403289	10
1399403509	9
1399403569	8
1399403619	7
1399403649	6
1399403669	5
1399403699	6
1399403909	7
1399404209	8
1399404219	9
1399404329	10
1399404339	11
1399404559	10
1399405762	9
1399406859	8
1399407499	7
1399407579	6
1399407709	7
1399408509	6
1399409079	5
1399409509	6
1399409589	5
1399411549	6
1399412519	5
1399412559	6
1399413209	7
1399414609	6
1399414649	7
1399415077	6
1399415229	7
1399415269	6
1399415439	4
1399416239	3
1399416540	2
1399416549	3
1399417810	2
1399419310	1
1399424600	0
1399436320	1
1399436420	0
1399441930	1
1399443040	0
1399445911	1
1399450972	2
1399452852	1
1399454002	2
1399455872	1
1399457882	0
1399458714	1
1399460613	2
1399460663	3
1399461673	2
1399461743	3
1399461783	2
1399461793	3
1399462313	4
1399462653	3
1399462663	4
1399463313	5
1399463803	4
1399463853	5
1399463873	6
1399464123	7
1399464183	6
1399464193	7
1399464293	6
1399465363	7
1399465483	6
1399465583	7
1399465893	6
1399466103	7
1399466183	8
1399466283	7
1399467203	6
1399469433	7
1399471274	6
1399471413	7
1399472063	8
1399472773	9
1399472783	10
1399473283	11
1399473543	12
1399473693	13
1399473773	12
1399474093	13
1399474103	14
1399474133	16
1399474163	17
1399474166	16
1399474173	17
1399474623	16
1399474863	15
1399475063	16
1399475313	15
1399475373	17
1399475504	18
1399475704	17
1399475714	18
1399475827	17
1399476065	18
1399476105	19
1399476165	18
1399476215	19
1399476455	18
1399476885	19
1399477255	18
1399477415	19
1399477565	20
1399477595	19
1399477685	20
1399477735	19
1399477785	18
1399477865	19
1399477915	18
1399477985	19
1399478075	20

cron.php
<?PHP

class Cron
{
	private static $instance = null;
	public static function GetInstance()
	{
		if (self::$instance == null)
			self::$instance = new Cron();
		return self::$instance;
	}

	private $schedules;
	private $airdrop_second = true;
	private $announce = 0;
	private $igtime = 0;

	private function __construct()
	{
		$this->schedules = Data::ReadCron();
		if ($this->schedules == null)
			$this->schedules = array(
				"cron_hours" => strtotime("next full hour"),
				"cron_airdrop" => time() + 30 * 60,
				"cron_announce" => time() + 5 * 60,
				"cron_playercount" => time() + 2 * 60,
				"cron_joinquit" => time() + 30,
				"cron_igtime" => time() + 60,
				"cron_vote" => time() + 10,
				"cron_parachute" => time() + 1,
			);
	}

	public function Run()
	{
		$new_schedules = array();
		$ctime = time();

		foreach ($this->schedules as $callback => $time)
			if ($ctime >= $time)
				$new_schedules[$callback] = $this->$callback();

		if (count($new_schedules) > 0)
		{
			$this->schedules = array_merge($this->schedules, $new_schedules);
			Data::WriteCrons($this->schedules);
		}
	}

	/**
	 * Full hour announce
	 */
	private function cron_hours()
	{
		echo "Announcing a passed hour.\n";
		RustRcon::GetInstance()->Send('say "' . Language::Text('hour', array('[color #FF0000]' . intval(date('G')) . ':00')) .'"');
		return strtotime('+1 hour');
	}

	/*
	 * Airdrop call
	 */
	private function cron_airdrop()
	{
		WaitForID(RustRcon::GetInstance()->Send('status'), 'cron_airdrop_response', $this);
		return time() + (30 * 60) + rand(0, 15 * 60);
	}

	public function cron_airdrop_response($response)
	{
		$players = Manager::GetInstance()->ParsePlayers($response);

		$drop = 0;
		if (count($players) > 1)
		{
			if (count($players) <= 4)
			{
				if ($this->airdrop_second)
					$drop = 1;
				$this->airdrop_second = !$this->airdrop_second;
			}
			elseif (count($players) < 20)
				$drop = 1;
			elseif (count($players) >= 20)
				$drop = 2;
		}

		for ($i = 0; $i < $drop; $i++)
			RustRcon::GetInstance()->Send('airdrop.drop');

		if ($drop > 0)
		{
			echo "Announcing airdrop.\n";
			RustRcon::GetInstance()->Send('say "' . Language::Text('airdrop') . '"');
		}
	}

	/**
	 * Cycling announces
	 */
	private function cron_announce()
	{
		$announces = Data::ReadAnnounces();
		if (++$this->announce >= count($announces))
			$this->announce = 0;

		$announce_content = $announces[$this->announce];

		echo "Announcing static messages.\n";
		foreach ($announce_content as $announce_text)
			RustRcon::GetInstance()->Send('say "[color #BBBBBB]' . $announce_text . '"');

		return time() + (1.5 * 60);
	}

	/**
	 * Player count announce
	 */
	private function cron_playercount()
	{
		WaitForID(RustRcon::GetInstance()->Send('status'), 'cmd_count_response', Commands::GetInstance());
		return time() + (2.25 * 60);
	}

	/**
	 * Join & Quit messages
	 */
	private function cron_joinquit()
	{
		Manager::GetInstance()->VipColors();
		Commands::GetInstance()->Timeout();
		WaitForID(RustRcon::GetInstance()->Send('status'), 'cron_joinquit_response', $this);
		$this->all_players = Data::ReadPlayers();
		return time() + 10;
	}

	private $prev_players = array();
	private $all_players = array();
	public function cron_joinquit_response($response)
	{
		$all_players = $this->all_players;
		$players = Manager::GetInstance()->ParsePlayers($response);
		$playerdata = Data::ReadPlayers();

		$test_for = (count($this->prev_players) == 0 ? Data::ReadPlayers() : $this->prev_players);

		$left = array_diff($this->prev_players, array_keys($players));

		$prev_players = $this->prev_players;
		if (count($prev_players) == 0)
			foreach ($all_players as $player => $data)
				if ($data['online'])
					$prev_players[] = $player;

		foreach ($players as $player => $data)
			if (isset($all_players[$player]))
				$all_players[$player]['last'] = time();
		Data::WritePlayers($all_players);

		$joined = array_diff(array_keys($players), $prev_players);
		$new = array_diff($joined, array_keys($all_players));
		$old = array_intersect($joined, array_keys($all_players));

/*		foreach ($new as $id => $maybe)
		{
			$test = Data::ReadPlayer($players[$maybe]['id']);
			if ($test != null && $test->first < time() - 60)
			{
				unset($new[$id]);
				$old[] = $maybe;
			}
		}
*/

		foreach ($players as $player => $data)
			if (!isset($playerdata[$player]))
				$playerdata[$player] = $data;

		$rcon = RustRcon::GetInstance();

/*		if (count($players) < 38 && true)
		{
		foreach ($left as $left_player)
		{
			echo "$left_player left.\n";
			IRC::WriteToIRC("#QUIT", $left_player);
			Data::WriteChatLine('#QUIT', $left_player);
			$rcon->Send('say "' . Language::Text('left', array('[color ' . $playerdata[$left_player]['color'] . ']' . $left_player)) . '"');
		}
		}
 */
		foreach ($joined as $joined_player)
		{
			Data::WriteVACCheck($players[$joined_player]['id']);
			$country = geoip_country_name_by_name($players[$joined_player]['ip']);

/*			if (count($players) < 38 && true)
			{
				echo "$joined_player joined from $country.\n";
				IRC::WriteToIRC("#JOIN", $joined_player . " from " . $country . " (" . (in_array($joined_player, $new) ? 'new' : 'old') . ")");
				Data::WriteChatLine('#JOIN', $joined_player);
				$rcon->Send('say "' . Language::Text('join', array('[color ' . (isset($playerdata[$joined_player]) ? $playerdata[$joined_player]['color'] : '#FF66FF') . ']' . $joined_player, '[color #FF0000]' . $country)) . '"');
			}
 */		

			if ($country == "Russian Federation")
			{
				$rcon->Send('banid "' . $players[$joined_player]['id'] . '" "No Russians allowed."');
				$rcon->Send('kick "' . $joined_player . '"');
				$rcon->Send('say "' . Language::Text('rusfag', array($joined_player)) . '"');
				Data::WritePlayer($players[$joined_player]);
			}

			if (in_array($country, array("Saudi Arabia", "United States", "Israel", "China", "Turkey", "Canada")))
			{
				echo "$country is flagged, will be kicked.\n";
				$rcon->Send('say "Kicking ' . $joined_player . ' for an insanely high ping."');
				IRC::WriteToIRC("#KICK", $joined_player . " (ping too high)");
				$rcon->Send('kick "' . $joined_player . '"');
			}
		}

/*		if (count($old) > 0)
			if (count($players) < 38)
			$rcon->Send('say "' . Language::Text('back', array(implode(' und ', $old))) . '"');
 */
		$karma = Data::ReadKarma();
		foreach ($joined as $joined_player)
		{
			$messages = Data::ReadMessages($joined_player);
			if (count($messages) > 0)
			{
				foreach ($messages as $message)
				{
					$rcon->Send('say "' . Language::Text('msg', array($message[1], $joined_player)) . '"');
					$rcon->Send('say "[color #BBBBBB]' . $message[2] . '"');
				}
			}
			if (isset($karma[$players[$joined_player]['id']]))
				if ($karma[$players[$joined_player]['id']] <= -5)
					if (count($players) < 38)
					RustRcon::GetInstance()->Send('say "' . Language::Text('karmawarn', array('[color ' . $playerdata[$joined_player]['color'] . ']' . $joined_player)) . '"');
		}
/*
		if (count($new) > 0)
		{
			foreach ($new as $player)
			{
				Data::WritePlayer($players[$player]);
			}
			$rcon->Send('say "' . Language::Text('welcome', array(implode(' und ', $new))) . '"');
			$rcon->Send('say "' . Language::Text('first') . '"');
		}
 */
		$this->prev_players = array_keys($players);
	}
	
	private function cron_vote()
	{
		Commands::GetInstance()->process_vote();
		return time() + 10;
	}
	
	/**
	 * In-game time
	 */
	private function cron_igtime()
	{
		WaitForID(RustRcon::GetInstance()->Send('env.time'), 'cron_igtime_response', $this);
		return time() + 20;
	}

	public function cron_igtime_response($response)
	{
		$time = floor(floatval(substr($response->Response(), 14)));
		if ($time != $this->igtime)
		{
			if ($time == 0 || $time == 6 || $time == 12 || $time == 18)
			{
				echo "Announcing $time:00 in-game.\n";
				RustRcon::GetInstance()->Send('say "' . Language::Text('inhour', array('[color #00FF00]' . $time . ':00')) . '"');
			}
		}
		$this->igtime = $time;
	}

	public function cron_parachute()
	{
		Commands::GetInstance()->TimeoutFalldmg();
		return time() + 1;
	}
}
